File: playwright-java/src/main/java/testtracks/diagnostics/SecretRedactor.java
package testtracks.diagnostics;

import java.util.regex.Pattern;

public final class SecretRedactor {
  private static final String REPLACEMENT = "[REDACTED]";
  private static final String JSON_SECRET_FIELDS =
      "accessToken|apiKey|authorization|clientSecret|cookie|password|refreshToken|secret|sessionId|token";
  private static final String KEY_VALUE_SECRET_NAMES =
      "access_token|api_key|client_secret|password|refresh_token|sessionid|token";
  private static final String HEADER_SECRET_NAMES = "Authorization|Cookie|Set-Cookie|X-Api-Key";
  private static final Pattern JSON_SECRET =
      Pattern.compile(
          "(\"(?:" + JSON_SECRET_FIELDS + ")\"\\s*:\\s*\")([^\"]*)(\")", Pattern.CASE_INSENSITIVE);
  private static final Pattern KEY_VALUE_SECRET =
      Pattern.compile("((?:" + KEY_VALUE_SECRET_NAMES + ")=)[^;\\s&]+", Pattern.CASE_INSENSITIVE);
  private static final Pattern HEADER_SECRET =
      Pattern.compile(
          "((?:" + HEADER_SECRET_NAMES + ")\\s*:\\s*)[^\\r\\n]+", Pattern.CASE_INSENSITIVE);

  private SecretRedactor() {}

  public static String redact(String value) {
    if (value == null || value.isBlank()) {
      return "";
    }

    var redacted = JSON_SECRET.matcher(value).replaceAll("$1" + REPLACEMENT + "$3");
    redacted = KEY_VALUE_SECRET.matcher(redacted).replaceAll("$1" + REPLACEMENT);
    return HEADER_SECRET.matcher(redacted).replaceAll("$1" + REPLACEMENT);
  }
}

File: playwright-csharp/src/Diagnostics/SecretRedactor.cs
using System.Text.RegularExpressions;

namespace TestTracks.Playwright.CSharp.Diagnostics;

/// <summary>
/// Removes common secrets from evidence before it is written to reports or failures.
/// </summary>
public static class SecretRedactor
{
    private const string Replacement = "[REDACTED]";
    private const RegexOptions Options = RegexOptions.IgnoreCase | RegexOptions.CultureInvariant;

    private static readonly string[] JsonSecretFields =
    [
        "accessToken",
        "apiKey",
        "authorization",
        "clientSecret",
        "cookie",
        "password",
        "refreshToken",
        "secret",
        "sessionId",
        "token"
    ];

    private static readonly string[] KeyValueSecretNames =
    [
        "access_token",
        "api_key",
        "client_secret",
        "password",
        "refresh_token",
        "sessionid",
        "token"
    ];

    private static readonly string[] HeaderSecretNames =
    [
        "Authorization",
        "Cookie",
        "Set-Cookie",
        "X-Api-Key"
    ];

    private static readonly string JsonSecretFieldPattern = string.Join("|", JsonSecretFields.Select(Regex.Escape));
    private static readonly string KeyValueSecretNamePattern = string.Join("|", KeyValueSecretNames.Select(Regex.Escape));
    private static readonly string HeaderSecretNamePattern = string.Join("|", HeaderSecretNames.Select(Regex.Escape));
    private static readonly Regex JsonSecretRegex = new(
        $"""("(?:{JsonSecretFieldPattern})"\s*:\s*")([^"]*)(")""",
        Options);
    private static readonly Regex KeyValueSecretRegex = new(
        $"""((?:{KeyValueSecretNamePattern})=)[^;\s&]+""",
        Options);
    private static readonly Regex HeaderSecretRegex = new(
        $"""((?:{HeaderSecretNamePattern})\s*:\s*)[^\r\n]+""",
        Options);

    /// <summary>
    /// Redacts common secret fields, key-value pairs and headers from text.
    /// </summary>
    public static string Redact(string? value)
    {
        if (string.IsNullOrEmpty(value))
        {
            return string.Empty;
        }

        var redacted = JsonSecretRegex.Replace(value, $"$1{Replacement}$3");
        redacted = KeyValueSecretRegex.Replace(redacted, $"$1{Replacement}");
        redacted = HeaderSecretRegex.Replace(redacted, $"$1{Replacement}");
        return redacted;
    }
}
